### PR TITLE
Read Go Build Flags from env `GO_BUILD_FLAGS` (#14396)

### LINKED ISSUE #14396: [makefile can't set GO_BUILD_FLAGS through env ]
### What happened?

Makefile
`
.PHONY: build
build:
	GO_BUILD_FLAGS="-v" ./scripts/build.sh
	./bin/etcd --version
	./bin/etcdctl version
	./bin/etcdutl version
`
GO_BUILD_FLAGS="-v"  will overwrite GO_BUILD_FLAGS in env.

### What did you expect to happen?

Makefile
`
.PHONY: build
build:
	GO_BUILD_FLAGS="-v" ./scripts/build.sh
	./bin/etcd --version
	./bin/etcdctl version
	./bin/etcdutl version
`
It should set as following.
` GO_BUILD_FLAGS +=" -v"  `

### How can we reproduce it (as minimally and precisely as possible)?

Repair it in Makefile!!!
`
.PHONY: build
build:
	GO_BUILD_FLAGS="-v" ./scripts/build.sh
	./bin/etcd --version
	./bin/etcdctl version
	./bin/etcdutl version
`
It should set as following.
` GO_BUILD_FLAGS +=" -v"  `

### Anything else we need to know?

_No response_

### Etcd version (please run commands below)

<details>

```console
$ etcd --version
# paste output here

$ etcdctl version
# paste output here
```

It affects all etcd versions obviously.

</details>

### Etcd configuration (command line flags or environment variables)

<details>

# paste your configuration here

</details>

### Etcd debug information (please run commands blow, feel free to obfuscate the IP address or FQDN in the output)

<details>

```console
$ etcdctl member list -w table
# paste output here

$ etcdctl --endpoints=<member list> endpoint status -w table
# paste output here
```

</details>

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Bhargav-InfraCloud**
# Motivation
[Issue][1] #14396

# Changes
1. `make build` will read `GO_BUILD_FLAGS` from env and prepend to go build flags.
2. ~~New make target `verify-go-build-flags-from-env` to runs tests from [build-test.sh][3].~~
3. ~~A test case `default-arch-build-flags-from-env` in [github actions][4] to runs tests from [build-test.sh][3] in pipeline.~~
    1. ~~New flag `upload-artifact` and condition to enable/disable upload of result JUnit XML selectively.~~
    2. ~~Syntax change of env export as per [this guide][5].~~
        ~~export JUNIT_REPORT_DIR=$(realpath ${TARGET})~~
        ~~# Changed to~~
        ~~echo "JUNIT_REPORT_DIR=$(realpath ${TARGET})" >> $GITHUB_ENV~~
4. A note about the change is added in [CONTRIBUTING.md][7].

# Summary
`make build` to read build flags from env `GO_BUILD_FLAGS` and prepend to those in [Makefile][2] for building executables (`etcd`, `etcdctl` & `etcdutl`). ~~The tests related to the same are added in [make verify][6] for local run and [github actions][4] for pipeline.~~

# Related Issues
Closes #14396 

[1]: https://github.com/etcd-io/etcd/issues/14396
[2]: https://github.com/etcd-io/etcd/blob/main/Makefile#L3
[3]: https://github.com/etcd-io/etcd/blob/main/scripts/build-test.sh
[4]: https://github.com/etcd-io/etcd/blob/main/.github/workflows/tests.yaml
[5]: https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable
[6]: https://github.com/etcd-io/etcd/blob/main/Makefile#L42
[7]: https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: scripts/build-test.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        echo " (matches expected: ${occurrences}, got: ${count})"
+
+        return 255
+    fi
+
+    echo "SUCCESS: Build flags with args from env GO_BUILD_FLAGS read successfully"
+    return 0
+}
+
+"$@"
```
Comment (ahrtr) [inline scripts/build-test.sh:56]: Please address "`no newline at end of file`".

File: scripts/build-test.sh [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    # Build command is directed to stdout for counting occurrences
+    # Order of flags should be same as that in env GO_BUILD_FLAGS (as prepended) + that in Makefile
+    local count 
+    count=$(GO_BUILD_FLAGS="${extra_flag} ${arg1}" make build 2>&1 | grep -c "'go' 'build' '${extra_flag}' '${arg1}' '-v'")
+
+    if [ "$count" -ne "$occurrences" ]; then
+        # Failure case: Number of occurrences doesn't match with expected
+        printf "Error: Build flags with args from env GO_BUILD_FLAGS not found in one or more logs"
+        echo " (matches expected: ${occurrences}, got: ${count})"
```
Comment (ahrtr) [inline scripts/build-test.sh:47]: Please reuse the existing log functions. FYI. https://github.com/etcd-io/etcd/blob/61e9ec9833917a3e470d410c9e9c1e40ec8340c8/scripts/test_lib.sh#L28-L62

**Comment by ahrtr** [comment]
It seems that it's a little over-engineering. I think we only need the changes in [CONTRIBUTING.md](https://github.com/etcd-io/etcd/pull/14868/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055) and line 3 of [Makefile](https://github.com/etcd-io/etcd/pull/14868/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52).

File: scripts/build-test.sh (hunk 1 above)
Comment (Bhargav-InfraCloud) [reply scripts/build-test.sh]: Done. Thanks!

File: scripts/build-test.sh (hunk 2 above)
Comment (Bhargav-InfraCloud) [reply scripts/build-test.sh]: scripts/build-test.sh.

**Comment by Bhargav-InfraCloud** [comment]
> It seems that it's a little over-engineering. I think we only need the changes in [CONTRIBUTING.md](https://github.com/etcd-io/etcd/pull/14868/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055) and line 3 of [Makefile](https://github.com/etcd-io/etcd/pull/14868/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52).

@ahrtr Agreed on over engineering part. But since it is a change in Makefile, and later for bugs in Makefile/shell are harder to debug (when compared with programming languages like Go), especially when shell doesn't throw errors on typos and `'foo' 'bar` and `'foo bar'` are considered differently while reading from flags. So, I tot there should be a check in place.

Helps in etcd's functionality - No :no_mouth:
Contributes to developer experience - Yes :grin:

That was my thought at least. Please confirm whether or not to have them and I'll update the branch accordingly.

**Comment by serathius** [comment]
Agree with @ahrtr Doesn't seem like a thing that we anyone would have a hard dependency and it should be easy to notice if it breaks.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -57,6 +57,8 @@ Setup environment:
   - For ubuntu and debian run `sudo apt-get install build-essentials`
 - Verify that everything is installed by running `make build`
 
+Note: In `make build`, the verbosity (`-v` flag) is enabled by default. Any additional go build flags, if required, should be added to env `GO_BUILD_FLAGS` before running `make build`.
```
Comment (serathius) [inline CONTRIBUTING.md:60]: Don't think this information is relevant enough to be put in contributing guide. I would prefer to keep them concise so new contributors will not be overwhelmed with information.

Comment (Bhargav-InfraCloud) [reply CONTRIBUTING.md:60]: Makes sense. I'll try simplifying it.

Comment (Bhargav-InfraCloud) [reply CONTRIBUTING.md]: Done. Please check now @serathius

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -57,6 +57,11 @@ Setup environment:
   - For ubuntu and debian run `sudo apt-get install build-essentials`
 - Verify that everything is installed by running `make build`
 
+Note: `make build` runs with `-v`. Other build flags can be added through env `GO_BUILD_FLAGS`, **if required**. Eg.,
+```console
+GO_BUILD_FLAGS="-tags linux" make build
```
Comment (ahrtr) [inline CONTRIBUTING.md:62]: `-tags linux` is a valid go build flag, but it isn't a good example, 

```suggestion
GO_BUILD_FLAGS="-buildmode=pie" make build
```

Comment (Bhargav-InfraCloud) [reply CONTRIBUTING.md]: Done. Thanks!

**Comment by Bhargav-InfraCloud** [comment]
Please check now. Thanks!

**Review by ahrtr — APPROVED** [review]
LGTM

Thank you @Bhargav-InfraCloud

**Review by serathius — APPROVED** [review]

**Comment by Bhargav-InfraCloud** [comment]
Thanks again @ahrtr @serathius 😊

**Pull request merged**